Fixes Issue #6580 Install to namespaces other than karmada-system via helm chart.

omsuneri opened the pull request:
**What type of PR is this?**
/kind bug

**What this PR does / why we need it**:
this PR fixes a issue where the karmada helm chart could only be installed in the `karmada-system` namespace previously, attempting to install karmada in any other namespace would result in controller crashes and certificate verification errors

**Which issue(s) this PR fixes**:
when users tried to install karmada using helm in a namespace other than `karmada-system`, they encountered errors

Fixes #6580 

the following chnages are tested and verified 
@zhzhuang-zju please review this PR and request if any further chnages are required !!

zhzhuang-zju commented:
> Commit sha: [24cadbe](https://github.com/karmada-io/karmada/pull/6601/commits/24cadbe9dc814beeb98bf908d7b66e6c07d624d4), Author: Om Santosh Suneri, Committer: GitHub; Expected "Om Santosh Suneri [142336291+omsuneri@users.noreply.github.com](mailto:142336291+omsuneri@users.noreply.github.com)", but got "omsuneri [omsuneri@gmail.com](mailto:omsuneri@gmail.com)".

I guess you have "Keep my email addresses private" enabled
<img width="718" height="42" alt="image" src="https://github.com/user-attachments/assets/03ae193b-0aeb-4844-adce-07f858cdd971" />

omsuneri committed:
helm: fix APIService namespace references for custom namespace

Signed-off-by: Om Santosh Suneri <142336291+omsuneri@users.noreply.github.com>

omsuneri committed:
improve systemNamespace parameter documentation

Signed-off-by: Om Santosh Suneri <142336291+omsuneri@users.noreply.github.com>

omsuneri commented:
@zhzhuang-zju currently some of the tests are failing with the chnages i m correcting all the chnages to follow the test !

omsuneri committed:
Fix template corruption in _karmada_apiservice.tpl

removed duplicate service definitions and fix incomplete APIService
template that was causing test failures.

Signed-off-by: Om Santosh Suneri <142336291+omsuneri@users.noreply.github.com>

zhzhuang-zju commented on charts/karmada/templates/_karmada_apiservice.tpl:
```diff
@@ -17,21 +16,9 @@ spec:
+    namespace: {{ include "karmada.namespace" . }}
   version: v1alpha1
   versionPriority: 10
----
-apiVersion: v1
-kind: Service
-metadata:
-  name: {{ $name }}-aggregated-apiserver
-  namespace: {{ $systemNamespace }}
-  labels:
-    app: {{ $name }}-aggregated-apiserver
-    {{- include "karmada.commonLabels" . | nindent 4 }}
-spec:
-  type: ExternalName
-  externalName: {{ $name }}-aggregated-apiserver.{{ include "karmada.namespace" . }}.svc.{{ .Values.clusterDomain }}
```
> the problem was that when karmada is installed in a custom namespace, these ExternalName services create a DNS resolution loop because the APIService points to a service in the same namespace where the ExternalName service exists, which then points back to itself, causing circular DNS resolution.

They do not cause DNS resolution loops. These externalName type Services are deployed under the karmada-apiserver context and are used to point to the same-named services deployed under the karmada-host context.

omsuneri replied:
thank you for the clarification @zhzhuang-zju  I understand now that I misunderstood the purpose of these ExternalName services and should not have deleted these ExternalName services this was my mistake 
Would you like me to revert the deletion of the ExternalName services and keep only the APIService namespace reference fixes ???

omsuneri commented:
> > test result :
> > before
> > `helm template karmada ./charts/karmada --namespace custom-name | grep "namespace:"`
> > returns: APIServices referenced karmada-system while services deployed to custom-ns - service discovery failed
> > after
> > `helm template karmada ./charts/karmada --namespace custom-name | grep "namespace:" `
> > returns: all namespace references point to custom-ns service discovery works
> 
> Great! Are there any tests for `helm install`?

Yes I've tested the actual helm install functionality and it works fine

omsuneri committed:
fix: support custom namespace for APIService references

Enable Karmada Helm chart installation in custom namespaces by fixing
APIService namespace references to use deployment namespace instead of
hardcoded karmada-system namespace.

Signed-off-by: Om Santosh Suneri <142336291+omsuneri@users.noreply.github.com>

RainbowMango commented:
Ping @zhzhuang-zju

zhzhuang-zju commented on charts/karmada/values.yaml:
```diff
@@ -41,6 +41,8 @@ installMode: "host"
 clusterDomain: "cluster.local"
 
+## @param systemNamespace the namespace used for system resources
+## This can be different from the namespace where Karmada components are deployed
```
+1, nice comment! And could be more detailed

```suggestion
## @param systemNamespace the namespace used for karmada resources, e.g. services, configmaps, etc.
## This can be different from the namespace where Karmada components are deployed (Release.Namespace).
```

zhzhuang-zju reviewed (commented):
After investigation, I found that the actual blocker for installing Karmada via Helm into a namespace other than `karmada-system` is related to certificates. When generating the certificates, the Subject Alternative Name (SAN) must include the DNS names for both `systemNamespace` and `release.Namespace`.
So, we can add a cert template in charts/karmada/templates/_helpers.tpl, like:
```md
{{- define "karmada.cert.hosts" -}}
{{- $hosts := .Values.certs.auto.hosts }}
{{- if and (.Values.systemNamespace) (ne .Values.systemNamespace .Release.Namespace) }}
{{- $hosts = append $hosts
  "*.{{ .Values.systemNamespace }}.svc.{{ .Values.clusterDomain }}"
-}}
{{- $hosts = append $hosts
  "*.{{ .Values.systemNamespace }}.svc"
-}}
{{- end }}
{{- $hosts | toJson }}
{{- end }}
```

And modify the code of pre-install-job.yaml.
https://github.com/karmada-io/karmada/blob/3ff3292b3db5e35d4d54216cbdf9f9b7cbf92352/charts/karmada/templates/pre-install-job.yaml#L446
```go
          echo '{"CN":"system:admin","hosts":{{ tpl (include "karmada.cert.hosts" .) . }},"names":[{"O":"system:masters"}],"key":{"algo":"rsa","size":{{ .Values.certs.auto.rsaSize }}}}' | cfssl gencert -ca=/opt/certs/server-ca.crt -ca-key=/opt/certs/server-ca.key -config=/opt/certs/server-ca-config.json - | cfssljson -bare /opt/certs/karmada
```

https://github.com/karmada-io/karmada/blob/3ff3292b3db5e35d4d54216cbdf9f9b7cbf92352/charts/karmada/templates/pre-install-job.yaml#L448

```go
          echo '{"CN":"front-proxy-client","hosts":{{ tpl (include "karmada.cert.hosts" .) . }},"names":[{"O":"system:masters"}],"key":{"algo":"rsa","size":{{ .Values.certs.auto.rsaSize }}}}' | cfssl gencert -ca=/opt/certs/front-proxy-ca.crt -ca-key=/opt/certs/front-proxy-ca.key -config=/opt/certs/front-proxy-ca-config.json - | cfssljson -bare /opt/certs/front-proxy-client
```

zhzhuang-zju commented:
test reports:
- deploy karmada in karmada-system namespace
```bash
$ helm install karmada -n karmada-system --create-namespace --dependency-update ./charts/karmada --set apiServer.hostNetwork=true,apiServer.serviceType=NodePort
W1205 16:55:58.430445  456923 warnings.go:70] spec.SessionAffinity is ignored for headless services
NAME: karmada
LAST DEPLOYED: Fri Dec  5 16:55:47 2025
NAMESPACE: karmada-system
STATUS: deployed
REVISION: 1
TEST SUITE: None
$ kubectl get pods --namespace karmada-system      
NAME                                               READY   STATUS    RESTARTS   AGE
etcd-0                                             1/1     Running   0          6m1s
karmada-aggregated-apiserver-67d4c46964-hlbqh      1/1     Running   0          6m1s
karmada-apiserver-78c548cb95-knrcd                 1/1     Running   0          6m1s
karmada-controller-manager-76c6444c5-l9p5f         1/1     Running   0          6m1s
karmada-kube-controller-manager-74f66dd89f-ssdw7   1/1     Running   0          6m1s
karmada-scheduler-c97c44df7-bbnh6                  1/1     Running   0          6m1s
karmada-webhook-87dc4c849-txsjj                    1/1     Running   0          6m1s
$ kubectl get --kubeconfig karmada-apiserver.config pp
NAME                CONFLICT-RESOLUTION   PRIORITY   AGE
nginx-propagation   Abort                 0          2m57s
$ kubectl get --kubeconfig karmada-apiserver.config rb
NAME               SCHEDULED   FULLYAPPLIED   AGE
nginx-deployment   False                      3m2s
$ kubectl get --kubeconfig karmada-apiserver.config deployments.apps 
NAME    READY   UP-TO-DATE   AVAILABLE   AGE
nginx   0/2     0            0           3m13s
```

- deploy karmada in test namespace 
```bash
$ helm install karmada -n test --create-namespace --dependency-update ./charts/karmada --set apiServer.hostNetwork=true,apiServer.serviceType=NodePort
W1205 17:08:14.183251  464429 warnings.go:70] spec.SessionAffinity is ignored for headless services
NAME: karmada
LAST DEPLOYED: Fri Dec  5 17:08:00 2025
NAMESPACE: test
STATUS: deployed
REVISION: 1
TEST SUITE: None
$ kubectl get pods --namespace test
NAME                                               READY   STATUS    RESTARTS   AGE
etcd-0                                             1/1     Running   0          5m19s
karmada-aggregated-apiserver-75df65f77-6lfp6       1/1     Running   0          5m19s
karmada-apiserver-84b789d44-bnlb4                  1/1     Running   0          5m19s
karmada-controller-manager-76c6444c5-vqfpr         1/1     Running   0          5m19s
karmada-kube-controller-manager-74f66dd89f-pb6jc   1/1     Running   0          5m19s
karmada-scheduler-c97c44df7-j5cxx                  1/1     Running   0          5m19s
karmada-webhook-87dc4c849-z6ckb                    1/1     Running   0          5m19s
$ $ kubectl get --kubeconfig karmada-apiserver.config pp
NAME                CONFLICT-RESOLUTION   PRIORITY   AGE
nginx-propagation   Abort                 0          2m40s
$ kubectl get --kubeconfig karmada-apiserver.config rb
NAME               SCHEDULED   FULLYAPPLIED   AGE
nginx-deployment   False                      2m45s
$ kubectl get --kubeconfig karmada-apiserver.config deployments.apps 
NAME    READY   UP-TO-DATE   AVAILABLE   AGE
nginx   0/2     0            0           2m12s
```

-  deploy karmada in test namespace and set systemNamespace to test1
```bash
$ helm install karmada -n test --create-namespace --dependency-update ./charts/karmada --set apiServer.hostNetwork=true,apiServer.serviceType=NodePort,systemNamespac=test1
W1205 17:19:08.858847  470527 warnings.go:70] spec.SessionAffinity is ignored for headless services
NAME: karmada
LAST DEPLOYED: Fri Dec  5 17:18:57 2025
NAMESPACE: test
STATUS: deployed
REVISION: 1
TEST SUITE: None
$ kubectl get pods --namespace test 
NAME                                               READY   STATUS      RESTARTS   AGE
etcd-0                                             1/1     Running     0          25s
karmada-aggregated-apiserver-75df65f77-p6ssd       1/1     Running     0          25s
karmada-apiserver-84b789d44-42z7j                  1/1     Running     0          25s
karmada-controller-manager-76c6444c5-9rdzq         1/1     Running     0          25s
karmada-kube-controller-manager-74f66dd89f-djw8h   1/1     Running     0          25s
karmada-scheduler-c97c44df7-rh5fr                  1/1     Running     0          24s
karmada-static-resource-fgt6h                      0/1     Completed   0          25s
karmada-webhook-87dc4c849-xvh89                    1/1     Running     0          25s
$ kubectl get --kubeconfig karmada-apiserver.config rb             
NAME               SCHEDULED   FULLYAPPLIED   AGE
nginx-deployment   False                      13s
$ kubectl get --kubeconfig karmada-apiserver.config pp
NAME                CONFLICT-RESOLUTION   PRIORITY   AGE
nginx-propagation   Abort                 0          20s
$ kubectl get --kubeconfig karmada-apiserver.config deployments.apps 
NAME    READY   UP-TO-DATE   AVAILABLE   AGE
nginx   0/2     0            0           21s
```
